File: Ruby_version/Napakalaki/lib/napakalaki.rb
# encoding: utf-8
#begin napakalaki.rb

require 'singleton'
require_relative 'card_dealer'         # Incluye monster y treasure
require_relative 'cultist_player'      # Incluye combat_result, dice, cultist y player

module NapakalakiGame
  # Clase Napakalaki que representa el control principal del juego
  class Napakalaki
    include Singleton
    
    # Attr / datos miembro
    # -------------------------------------------------------
    
    #attr_reader :currentPlayer       # Jugador con el turno actual
    #attr_reader :currentMonster      # Monstruo en el turno actual
    #attr_reader :players             # Todos los jugadores
    #attr_reader :dealer              # El repartidor de cartas
    
    # -------------------------------------------------------
    
    # Constructor
    # -------------------------------------------------------

    def initialize
      @currentPlayer = nil
      @players = []
      @currentMonster = nil
      @dealer = CardDealer.instance
    end
    
    # -------------------------------------------------------
    
    # Consultores
    # -------------------------------------------------------
    
    def getCurrentPlayer
      @currentPlayer
    end

    def getCurrentMonster
      @currentMonster
    end
    
    # -------------------------------------------------------
    
    # Métodos
    # -------------------------------------------------------

    # Crea todos los jugadores con los nombres pasados en el array
    def initPlayers(names)
      names.each { |name| @players << Player.newPlayer(name) }
    end

    # Devuelve y actualiza el jugador al que le toca jugar
    def nextPlayer
      if @currentPlayer == nil
        @currentPlayer = @players.sample
      elsif @currentPlayer == @players.last
        @currentPlayer = @players.first
      else
        @currentPlayer = @players[@players.index(@currentPlayer) + 1]
      end
    end

    # Si se puede pasar al siguiente turno
    def nextTurnAllowed
      @currentPlayer == nil or @currentPlayer.validState
    end

    # Establece aleatoriamente los archienemigos de los jugadores
    def setEnemies
      @players.each do |player|
        enemy = @players.sample
        while player == enemy do
          enemy = @players.sample
        end
        player.setEnemy(enemy)
      end
    end

    # Desarrolla el combate
    def developCombat
      combatResult = @currentPlayer.combat(@currentMonster)
      if combatResult == CombatResult::LOSEANDCONVERT
        cultistCard = @dealer.nextCultist
        newCultist = CultistPlayer.new(@currentPlayer, cultistCard)
        @players.each do |player|
          if player.getName == @currentPlayer.name
            player.setEnemy(newCultist)
          end
        end
        index = @players.index(@currentPlayer)
        @currentPlayer = newCultist
        @players[index] = newCultist
      end
      @dealer.giveMonsterBack(@currentMonster)
      combatResult
    end

    # Descarta los tesoros visibles del jugador
    def discardVisibleTreasures(treasures)
      treasures.each do |treasure|
        @currentPlayer.discardVisibleTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      end
    end

    # Descarta los tesoros ocultos del jugador
    def discardHiddenTreasures(treasures)
      treasures.each do |treasure|
        @currentPlayer.discardHiddenTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      end
    end

    # Hace visible todos los tesoros que se pasan
    def makeTreasuresVisible(treasures)
      treasures.each { |t| @currentPlayer.makeTreasureVisible(t) }
    end

    # Inicia el juego
    def initGame(players)
      initPlayers(players)
      setEnemies
      @dealer.initCards
      nextTurn
    end

    # Pasa al siguiente turno
    def nextTurn
      stateOK = nextTurnAllowed
      if stateOK
        @currentMonster = @dealer.nextMonster
        @currentPlayer = nextPlayer
        if @currentPlayer.isDead
          @currentPlayer.initTreasures
        end
      end
      stateOK
    end

    # Si ha acabado la partida
    def endOfGame(result)
      result == CombatResult::WINGAME
    end

    # -------------------------------------------------------
    
    # Métodos privados
    # -------------------------------------------------------

    private :initPlayers
    private :nextPlayer
    private :nextTurnAllowed
    private :setEnemies
    
    # -------------------------------------------------------
  end
end

# end napakalaki.rb

File: Java_version/Napakalaki/src/napakalaki/Napakalaki.java
// begin Napakalaki.java
package napakalaki;

import java.util.ArrayList;
import java.util.Random;
import java.util.Iterator;

// Representación del juego
public class Napakalaki {
    
   private static final Napakalaki instance = new Napakalaki();
   private Player currentPlayer = null;
   private int indexCurrentPlayer = -1; //No aparece en el diagrama, para el método 
                                   //nextPlayer, comentado con el profesor.
   private ArrayList<Player> players = new ArrayList();
   private CardDealer dealer = CardDealer.getInstance();
   private Monster currentMonster;
   
   private void initPlayers(ArrayList<String> names) {
       for (String name : names) {
           players.add(new Player(name));
       }
   }
   
   //Podríamos implementar esto con un nuevo atributo de instance
   private Player nextPlayer() {
       if (indexCurrentPlayer == -1) {
           indexCurrentPlayer = new Random().nextInt(players.size());
       } else {
           indexCurrentPlayer++;
           if (indexCurrentPlayer == players.size()) 
               indexCurrentPlayer = 0;
       }
       currentPlayer = players.get(indexCurrentPlayer);
       return currentPlayer;
   }
   
   private boolean nextTurnAllowed() {
       boolean result;
       if (currentPlayer == null)
           result = true;
       else
           result = currentPlayer.validState();
       return result;
   }
   
   private void setEnemies() {
       Random rand = new Random();
       Player enemy;
       for (Player p : players) {
            do {
               enemy = players.get(rand.nextInt(players.size()));
            } while (p == enemy);
           p.setEnemy(enemy);
       }
   }
   
   public static Napakalaki getInstance() {
       return instance;
   }
   
   public CombatResult developCombat() {
       CombatResult cr = currentPlayer.combat(currentMonster);
       if (cr == CombatResult.LOSEANDCONVERT) {
            CultistPlayer sectario = new CultistPlayer (currentPlayer,
                dealer.nextCultist());
            players.set(players.indexOf(currentPlayer), sectario);
            for (int i = 0; i < players.size(); i++) {
                if (players.get(i).getEnemy() == currentPlayer)
                    players.get(i).setEnemy(sectario);
            }
            currentPlayer = sectario;
       }
       dealer.giveMonsterBack(currentMonster);
       return cr;
   }
   
   public void discardVisibleTreasures(ArrayList<Treasure> treasures) {
       Iterator <Treasure> i = treasures.iterator();
       Treasure t;
       while (i.hasNext()) {
           t = i.next();
           currentPlayer.discardVisibleTreasure(t);
           dealer.giveTreasureBack(t);
       }
   }
   
   public void discardHiddenTreasures(ArrayList<Treasure> treasures) {
       
   }
   
   public void makeTreasuresVisible(ArrayList<Treasure> treasures) {
       for (Treasure t : treasures)
           currentPlayer.makeTreasureVisible(t);
   }
   
   public void initGame(ArrayList<String> players) {
       initPlayers(players);
       setEnemies();
       dealer.initCards();
       nextTurn();
   }
   
   public Player getCurrentPlayer() {
       return currentPlayer;
   }
   
   public Monster getCurrentMonster() {
       return currentMonster;
   }
   
   public boolean nextTurn() {
       boolean stateOK = nextTurnAllowed();
       if (stateOK) {
           currentMonster = dealer.nextMonster();
           currentPlayer = nextPlayer();
           if (currentPlayer.isDead()) {
               currentPlayer.initTreasures();
           }
       }
       return stateOK;
   }
   
   public boolean endOfGame(CombatResult result) {
       return result == CombatResult.WINGAME;
   }
   
}

// end Napakalaki.java
